File: include/extras-kcoreaddons/libkshell.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const kshell_enums = enums;
const std = @import("std");

/// https://api-staging.kde.org/kshell.html
pub const kshell = struct {
    /// [Qt documentation](https://api-staging.kde.org/kshell.html#splitArgs)
    ///
    /// ``` param1: []const u8, param2: flag of kshell_enums.Option, param3: *kshell_enums.Errors, allocator: std.mem.Allocator ```
    pub fn SplitArgs(param1: []const u8, param2: i64, param3: *i32, allocator: std.mem.Allocator) [][]const u8 {
        const param1_str = qtc.libqt_string{
            .len = param1.len,
            .data = param1.ptr,
        };
        const _arr: qtc.libqt_list = qtc.KShell_SplitArgs(param1_str, @intCast(param2), @ptrCast(param3));
        const _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i| {
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]const u8, _arr.len) catch @panic("kshell.SplitArgs: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data = _str[i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("kshell.SplitArgs: Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// [Qt documentation](https://api-staging.kde.org/kshell.html#joinArgs)
    ///
    /// ``` param1: [][]const u8, allocator: std.mem.Allocator ```
    pub fn JoinArgs(param1: [][]const u8, allocator: std.mem.Allocator) []const u8 {
        var param1_arr = allocator.alloc(qtc.libqt_string, param1.len) catch @panic("kshell.JoinArgs: Memory allocation failed");
        defer allocator.free(param1_arr);
        for (param1, 0..param1.len) |item, i| {
            param1_arr[i] = .{
                .len = item.len,
                .data = item.ptr,
            };
        }
        const param1_list = qtc.libqt_list{
            .len = param1.len,
            .data = param1_arr.ptr,
        };
        const _str = qtc.KShell_JoinArgs(param1_list);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kshell.JoinArgs: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api-staging.kde.org/kshell.html#quoteArg)
    ///
    /// ``` param1: []const u8, allocator: std.mem.Allocator ```
    pub fn QuoteArg(param1: []const u8, allocator: std.mem.Allocator) []const u8 {
        const param1_str = qtc.libqt_string{
            .len = param1.len,
            .data = param1.ptr,
        };
        const _str = qtc.KShell_QuoteArg(param1_str);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kshell.QuoteArg: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api-staging.kde.org/kshell.html#tildeExpand)
    ///
    /// ``` param1: []const u8, allocator: std.mem.Allocator ```
    pub fn TildeExpand(param1: []const u8, allocator: std.mem.Allocator) []const u8 {
        const param1_str = qtc.libqt_string{
            .len = param1.len,
            .data = param1.ptr,
        };
        const _str = qtc.KShell_TildeExpand(param1_str);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kshell.TildeExpand: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api-staging.kde.org/kshell.html#tildeCollapse)
    ///
    /// ``` param1: []const u8, allocator: std.mem.Allocator ```
    pub fn TildeCollapse(param1: []const u8, allocator: std.mem.Allocator) []const u8 {
        const param1_str = qtc.libqt_string{
            .len = param1.len,
            .data = param1.ptr,
        };
        const _str = qtc.KShell_TildeCollapse(param1_str);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kshell.TildeCollapse: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }
};

/// https://api-staging.kde.org/kshell.html#types
pub const enums = struct {
    pub const Option = enum {
        pub const NoOptions: i32 = 0;
        pub const TildeExpand: i32 = 1;
        pub const AbortOnMeta: i32 = 2;
    };

    pub const Errors = enum {
        pub const NoError: i32 = 0;
        pub const BadQuoting: i32 = 1;
        pub const FoundMeta: i32 = 2;
    };
};

File: src/extras-kcoreaddons/libkshell.cpp
#include <KShell>
#include <QList>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <kshell.h>
#include "libkshell.h"
#include "libkshell.hxx"

libqt_list /* of libqt_string */ KShell_SplitArgs(const libqt_string param1, int param2, int* param3) {
    QString param1_QString = QString::fromUtf8(param1.data, param1.len);
    QList<QString> _ret = KShell::splitArgs(param1_QString, static_cast<KShell::Options>(param2), reinterpret_cast<KShell::Errors*>(param3));
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        QString _lv_ret = _ret[i];
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _lv_b = _lv_ret.toUtf8();
        libqt_string _lv_str;
        _lv_str.len = _lv_b.length();
        _lv_str.data = static_cast<const char*>(malloc(_lv_str.len + 1));
        memcpy((void*)_lv_str.data, _lv_b.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_string KShell_JoinArgs(const libqt_list /* of libqt_string */ param1) {
    QList<QString> param1_QList;
    param1_QList.reserve(param1.len);
    libqt_string* param1_arr = static_cast<libqt_string*>(param1.data);
    for (size_t i = 0; i < param1.len; ++i) {
        QString param1_arr_i_QString = QString::fromUtf8(param1_arr[i].data, param1_arr[i].len);
        param1_QList.push_back(param1_arr_i_QString);
    }
    QString _ret = KShell::joinArgs(param1_QList);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string KShell_QuoteArg(const libqt_string param1) {
    QString param1_QString = QString::fromUtf8(param1.data, param1.len);
    QString _ret = KShell::quoteArg(param1_QString);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string KShell_TildeExpand(const libqt_string param1) {
    QString param1_QString = QString::fromUtf8(param1.data, param1.len);
    QString _ret = KShell::tildeExpand(param1_QString);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string KShell_TildeCollapse(const libqt_string param1) {
    QString param1_QString = QString::fromUtf8(param1.data, param1.len);
    QString _ret = KShell::tildeCollapse(param1_QString);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}
